rmf_fleet_adapter/src/rmf_fleet_adapter/events/PerformAction.cpp

```diff
@@ -0,0 +1,248 @@
+  const std::string& category,
+  nlohmann::json desc,
+  rmf_traffic::Duration time_estimate,
+  rmf_task::events::SimpleEventStatePtr state,
+  std::function<void()> update,
+  std::function<void()> finished) -> std::shared_ptr<Active>
+{
+  auto active = std::make_shared<Active>(
+    Active(std::move(category), std::move(desc), time_estimate));
+  active->_assign_id = id;
+  active->_context = std::move(context);
+  active->_update = std::move(update);
+  active->_finished = std::move(finished);
+  active->_state = std::move(state);
+  active->_execute_action();
+  active->_expected_finish_time =
+    active->_context->now() + time_estimate;
+  return active;
+}
+
+//==============================================================================
+PerformAction::Active::Active(
+  const std::string& category,
+  nlohmann::json desc,
+  rmf_traffic::Duration time_estimate)
+: _action_category(std::move(category)),
+  _action_description(std::move(desc)),
+  _time_estimate(std::move(time_estimate))
+{
+  // Do nothing
+}
+
+//==============================================================================
+auto PerformAction::Active::state() const -> ConstStatePtr
+{
+  return _state;
+}
+
+//==============================================================================
+rmf_traffic::Duration PerformAction::Active::remaining_time_estimate() const
+{
+  if (_context->action_remaining_time().has_value())
+  {
+    return _context->action_remaining_time().value();
+  }
+
+  // If an estimate is not provided we compute one based on the expected finish
+  // time
+  return _expected_finish_time - _context->now();
+}
+
+//==============================================================================
+auto PerformAction::Active::backup() const -> Backup
+{
+  // PerformAction doesn't need to be backed up
+  return Backup::make(0, nlohmann::json());
+}
+
+//==============================================================================
+auto PerformAction::Active::interrupt(
+  std::function<void()> task_is_interrupted) -> Resume
+{
+  _state->update_status(Status::Standby);
+  _state->update_log().info("Going into standby for an interruption");
+  _state->update_dependencies({});
+
+  _context->worker().schedule(
+    [task_is_interrupted](const auto&)
+    {
+      task_is_interrupted();
+    });
+
+  // TODO(YV): Currently we will ask the robot to perform the action again when
+  // resumed. Future work can be to receive a callback from RobotUpdateHandle
+  // which can be used to ask the robot to resume from its interrupted state.
+  return Resume::make(
+    [w = weak_from_this()]()
+    {
+      if (const auto self = w.lock())
+      {
+        self->_execute_action();
+      }
+    });
+}
+
+//==============================================================================
+void PerformAction::Active::cancel()
+{
+  _state->update_status(Status::Canceled);
+  _state->update_log().info("Received signal to cancel");
+  _finished();
```

Comment:
We should think about ways to signal to the `ActionExecution` that it has been cancelled. Maybe give it a `bool okay() const` function, and set the `schedule` to `nullptr`? Maybe have it accept a callback that we can trigger when a cancellation happens?

Reply:
Done https://github.com/open-rmf/rmf_ros2/pull/155/commits/9a8ad648fa3b2baa07c273ea4ccfc5d941576c29